File: service_manager.go
package main

// ServiceManager abstracts OS-level process and service control so that
// the core logic can be tested without executing real system commands.
type ServiceManager interface {
	// StopService stops a named OS service (Windows service / systemd unit).
	StopService(id string) error
	// StartService starts a named OS service.
	StartService(id string) error
	// KillProcess terminates all running instances of the given executable name.
	KillProcess(exe string) error
	// RunCommand executes an arbitrary command with the given arguments.
	RunCommand(name string, args ...string) error
}

// StopAll stops every enabled service in cfg.Services using the provided manager.
func StopAll(m ServiceManager, cfg Config) []ActionResult {
	results := make([]ActionResult, 0)
	for _, svc := range cfg.Services {
		if !svc.Enabled {
			continue
		}
		err := m.StopService(svc.ID)
		results = append(results, ActionResult{Name: svc.Name, Err: err})
	}
	return results
}

// StartAll starts every enabled service in cfg.Services.
func StartAll(m ServiceManager, cfg Config) []ActionResult {
	results := make([]ActionResult, 0)
	for _, svc := range cfg.Services {
		if !svc.Enabled {
			continue
		}
		err := m.StartService(svc.ID)
		results = append(results, ActionResult{Name: svc.Name, Err: err})
	}
	return results
}

// KillAll kills every executable in every enabled ProcessEntry.
func KillAll(m ServiceManager, cfg Config) []ActionResult {
	results := make([]ActionResult, 0)
	for _, proc := range cfg.Processes {
		if !proc.Enabled {
			continue
		}
		for _, exe := range proc.Exes {
			err := m.KillProcess(exe)
			results = append(results, ActionResult{Name: proc.Name + " (" + exe + ")", Err: err})
		}
	}
	return results
}

// ActionResult captures the outcome of a single stop/start/kill operation.
type ActionResult struct {
	Name string
	Err  error
}

// OK returns true when the action succeeded.
func (r ActionResult) OK() bool { return r.Err == nil }

File: src/service-manager.js
'use strict';

const { execSync } = require('child_process');

/**
 * Determines the current platform.
 * @returns {'win32'|'linux'|'darwin'|string}
 */
function getPlatform() {
  return process.platform;
}

/**
 * Stop a Windows service or Linux systemd unit.
 * @param {object} service - service descriptor from config
 * @returns {{ success: boolean, message: string }}
 */
function stopService(service) {
  try {
    if (getPlatform() === 'win32') {
      if (service.type === 'windows-service') {
        execSync(`net stop "${service.id}"`, { stdio: 'pipe' });
      } else if (service.type === 'process') {
        execSync(`taskkill /IM "${service.id}.exe" /F`, { stdio: 'pipe' });
      }
    } else {
      // Linux / macOS – use systemctl or kill
      if (service.type === 'windows-service') {
        execSync(`systemctl stop "${service.id}"`, { stdio: 'pipe' });
      } else if (service.type === 'process') {
        execSync(`pkill -f "${service.id}" || true`, { stdio: 'pipe' });
      }
    }
    return { success: true, message: `${service.name} stopped.` };
  } catch (err) {
    return { success: false, message: `Failed to stop ${service.name}: ${err.message}` };
  }
}

/**
 * Start a Windows service or Linux systemd unit.
 * @param {object} service - service descriptor from config
 * @returns {{ success: boolean, message: string }}
 */
function startService(service) {
  try {
    if (getPlatform() === 'win32') {
      if (service.type === 'windows-service') {
        execSync(`net start "${service.id}"`, { stdio: 'pipe' });
      } else if (service.type === 'process') {
        execSync(`start "" "${service.id}.exe"`, { stdio: 'pipe' });
      }
    } else {
      if (service.type === 'windows-service') {
        execSync(`systemctl start "${service.id}"`, { stdio: 'pipe' });
      } else if (service.type === 'process') {
        execSync(`${service.id} &`, { stdio: 'pipe', shell: true });
      }
    }
    return { success: true, message: `${service.name} started.` };
  } catch (err) {
    return { success: false, message: `Failed to start ${service.name}: ${err.message}` };
  }
}

/**
 * Stop all enabled services in the config list.
 * @param {Array} services
 * @returns {Array<{ name: string, success: boolean, message: string }>}
 */
function stopAll(services) {
  return services
    .filter(s => s.enabled)
    .map(s => ({ name: s.name, ...stopService(s) }));
}

/**
 * Start all enabled services in the config list.
 * @param {Array} services
 * @returns {Array<{ name: string, success: boolean, message: string }>}
 */
function startAll(services) {
  return services
    .filter(s => s.enabled)
    .map(s => ({ name: s.name, ...startService(s) }));
}

module.exports = { stopService, startService, stopAll, startAll };
